### ISSUE TITLE
Support macros

### ISSUE THREAD

**Issue by miekg** [miekg/cf#14]
Support cfengine macro's so at least the std lib of cfengine3 parses completely.

**Comment by miekg** [comment]
enough support added to make it parse, but output can be better and these macros can be put at more places than we currently allow

**Comment by miekg** [comment]
from the unparseable directory:
~~~
bundle agent extractor
{
@if minimum_version(3.16)
  # Implementation for 3.16+
  vars:
    "container"
      data => classfiltercsv(something);
@else
  # Implementation for versions before 3.16
  vars:
    "container"
      data => readcsv(something);
@endif
}
~~~
the macro `@else` and `@endif` are not "seen" by the parser because we don't check for them in a constraint (and selection prolly).
If we add `Macro(b)` code to the parser at the right places I can get as far as:

~~~
    "container"
      data => readcsv(something)@endif
;
}
~~~
I.e. that would be with this patch
~~~ diff
diff --git internal/parse/parse.go internal/parse/parse.go
index b2dbd54..152a235 100644
--- internal/parse/parse.go
+++ internal/parse/parse.go
@@ -65,6 +65,7 @@ func BundleBody(b *rd.Builder) (ok bool) {
        defer b.Exit(&ok)
 
        Comments(b)
+       Macro(b)
 More:
        // Zero or more promiseguards (single : ) and then zero more classpromises.
        PromiseGuard(b)
@@ -136,7 +137,7 @@ func Promise(b *rd.Builder) (ok bool) {
        Comments(b)
 
        // zero or more constraints, and then a closing ;
-       return PromiseConstraints(b) && Match(b, chroma.Token{Type: chroma.Punctuation, Value: ";"})
+       return PromiseConstraints(b) && Match(b, chroma.Token{Type: chroma.Punctuation, Value: ";"}) && Macro(b)
 }
 
 func PromiseConstraints(b *rd.Builder) (ok bool) {
~~~

The problem here is that the Macro is a child of the constraint while in reality it's a child of the specification. The current tree impl I use, doesn't have parent pointers, so I can't climb up the tree easily

**Comment by miekg** [comment]
closed by #41
